File: java-implementation/src/main/java/com/inventory/products/guitars/Acoustic.java
package com.inventory.products.guitars;

// Represents an acoustic guitar as a concrete subclass of Guitar.
public class Acoustic extends Guitars {
    private boolean hasPickup; // Indicates if the acoustic guitar has a pickup

    // Constructs an instance with specified details
    public Acoustic(int id, String name, double price, int volume, String material, int stringCount, boolean hasPickup, String supplier) {
        super(id, name, price, volume, material, stringCount, supplier);
        this.hasPickup = hasPickup;
    }

    // Calculates the discount based on volume and month
    @Override
    public double calculateDiscount(int volume, int month) {
        if (month == 6 && volume > 3) {
            return price * volume * 0.50;
        }
        return 0;
    }

    // Provides a string representation of the acoustic guitar's details
    @Override
    public String displayProductInfo() {
        return "ID: " + id + "\nAcoustic Guitar: " + name + "\nMaterial: " + material
               + "\nHas Pickup: " + (hasPickup ? "Yes" : "No") + "\nPrice: €" + price
               + "\nString Count: " + stringCount + "\nVolume: " + volume + " units\nSupplier: " + supplier + "\n\n";
    }
}

File: cpp-implementation/src/products/guitars/Acoustic.cpp
#include "Product.ih"

// Constructor implementation for Acoustic guitars
Acoustic::Acoustic(int id, string name, double price, int volume, 
                    string material, int stringCount, bool hasPickup, 
                    string supplier)
    : Guitars(id, name, price, volume, material, stringCount, supplier), 
      hasPickup(hasPickup) {}

// Calculates the discount depending on the current month and the packaged quantity
double Acoustic::calculateDiscount(int volume, int month) const {
    if (month == 6 && volume > 3) {
        return price * volume * 0.50;
    }
    return 0;
}

// Displays all the information
string Acoustic::displayProductInfo() const {
    ostringstream info;
    info << "ID: " << id << "\nAcoustic Guitar: " << name << "\nSupplier: " << getSupplier()
         << "\nMaterial: " << material << "\nHas Pickup: " << (hasPickup ? "Yes" : "No")
         << "\nPrice: €" << price << "\nString Count: " << stringCount
         << "\nVolume: " << volume << " units\n\n";
    return info.str();
}
